Add support for preview-fn

ivarref opened the pull request:
Hello

And thanks for a great project!

I added support for `:preview-fn`, which I thought could be a neat and handy feature.

It's implemented using a simple bb script that will be used as the preview command.
The bb script will connect to 127.0.0.1 : 'preview-fn-server-port', send the selected item and print out the response.

Example usage:
```clojure
(require '[fzf.core :refer [fzf]])

(println (fzf {:preview-fn (fn [nam] (str "Hello, " nam " :-)"))} ["one" "two" "three"]))
```

That will display as:
![image](https://github.com/joakimen/fzf.clj/assets/74075/0794aa17-eb89-45a0-8389-219ab0b1bd96)

It appears I've messed up some whitespace, sorry about that. If that is a problem, let me know and I can (probably!) fix it.

Kind regards.

ivarref committed:
Add support for preview-fn

joakimen commented:
Hi, thanks for the contribution!

I really like this idea, it would make preview-functions testable and REPL-friendly.

A couple of things:

1. This implementation breaks the default usage of `core.fzf`. When `fzf` is called without a map containing the `:preview-fn` key as the first argument (e.g. `(fzf ["test"])` or `(fzf {:multi true} ["test"])`, `impl.fzf` still attempts to start the preview-server and throws

    ```clojure
    java.lang.IllegalArgumentException: No matching clause: java.io.Closeable fzf.core /src/fzf/core.clj:48:4
    ```
    
    `impl/fzf` must be updated so `:preview-fn` plays well with existing options
    
2. Since `:preview-fn` overwrites `:preview`, there should probably be a note about this in the `core/fzf` docstring, and perhaps throw an exception if both `:preview` and `:preview-fn` are specified
3. `impl/parse-opts` should handle the resolution of `:preview`/`:preview-fn` and the return of the `--preview` with the inline-script. This function should avoid doing IO to keep it easy to test, then we can do the IO (starting the server etc) in `impl/fzf` or a new, separate function
4. We need tests in `impl_test` that proves the behaviour. Some ideas that come to mind:
    1. prove that setting `:preview-fn` overwrites `:preview`
    2. prove that setting `:preview` but `:preview-fn` is nil/blank/missing leaves `:preview` intact

ivarref committed:
Address feedback

ivarref commented:
Hi, and thanks for the feedback!

I've pushed code and tests that addresses the feedback:

1. _plays well with existing options_: Fixed.
2. Both docstring and throwing if both are specified is added.
3. Good idea. Test added.
4. With the throwing behaviour we don't need to worry about this. Existing test using `:preview` still passes.

What do you think?

Regards.

joakimen committed:
Delegate more validation to clojure.spec

joakimen commented:
Looked over the changes and did some tests, seems great 💯 

Got some tips in the [Clojurian-slack](https://clojurians.slack.com/archives/C03S1KBA2/p1696417189164679) on how to delegate the validation of the preview-keys to clojure.spec, so updated that in `fzf/core` and updated the test accordingly.

Thanks again for the cool feature!
